### DISCUSSION TITLE
return value에 관한 질문

### DISCUSSION THREAD

**Discussion by Hsg2311** [MyeongWoonJang/cpp-study#87] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/75879925/200316566-758adfd8-ca35-4cdf-bcd7-724bf902d0df.png)

**Comment by Woon-2** [comment, marked as answer]
`ref`라는 `int&` `type`의 `parameter`에 `a`라는 `argument`가 전달됐습니다.   
그리고 그걸 다시 `return`하고 있네요.

9번 라인에서 `b`는 `a`의 `reference`가 되었습니다.

내부적으론 이런 일이 일어납니다.

`func`가 `call`되면서 `return value`를 임시적으로 저장할 메모리 공간이 `stack` 메모리에 확보됩니다.   
그리고 `func`의 끝 부분에서 그 메모리 공간에 `return value`를 `write`합니다.   
그 임시 메모리는 `func`를 `call`한 부분에 전달됩니다.

스택 메모리의 개요를 나타내면 다음과 같습니다.    
(완전히 엄밀하진 않습니다.)

### func call 전

`main을 위한 메모리`

### func call

`func를 위한 메모리` - pushed   
`func의 제일 첫 부분에 해당하는 명령어 주소` - pushed   
`func를 호출한 명령어의 주소` - pushed   
`return value를 위한 임시 메모리` - pushed   
`main을 위한 메모리`

### after call

`위에 있던 것들` - poped   
`return value를 위한 임시 메모리` - will be poped (명령어가 끝나고 나면 pop될 것입니다.)   
`main을 위한 메모리`

**Reply by Woon-2** [reply]
`return value`와 같이 `temporary object`로 존재하는 것들은, `code` 상에서 단 한 줄 동안만 생존합니다.

그래서 `return value`를 나중에도 쓰려면, `variable`에 저장해야 하죠.   

`return value를 위한 임시 메모리`는 `main을 위한 메모리`가 아니니까,   
`main을 위한 메모리`에 있는 `local variable`에 `return value(temporary object)`를 저장하는 겁니다.
